src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -149,45 +193,116 @@ internal static SafeSslHandle AllocateSslContext(SslProtocols protocols, SafeX50
                     SetSslCertificate(innerContext, certHandle!, certKeyHandle!);
                 }
 
-                if (sslAuthenticationOptions.IsServer && sslAuthenticationOptions.RemoteCertRequired)
+            }
+            catch
+            {
+                innerContext.Dispose();
+                throw;
+            }
+
+            return innerContext;
+        }
+
+        // This essentially wraps SSL* SSL_new()
+        internal static SafeSslHandle AllocateSslHandle(SafeFreeSslCredentials credential, SslAuthenticationOptions sslAuthenticationOptions)
+        {
+            SafeSslHandle? context = null;
+            SafeSslContextHandle? sslCtx = null;
+            SafeSslContextHandle? innerContext = null;
+            SslProtocols protocols = sslAuthenticationOptions.EnabledSslProtocols;
+            bool cacheSslContext = !DisableTlsResume && sslAuthenticationOptions.EncryptionPolicy == EncryptionPolicy.RequireEncryption &&
+                     sslAuthenticationOptions.CipherSuitesPolicy == null &&
+                    (!sslAuthenticationOptions.IsServer ||
+                    (sslAuthenticationOptions.ApplicationProtocols != null && sslAuthenticationOptions.ApplicationProtocols.Count != 0));
+
+            if (!Interop.Ssl.Tls13Supported)
+            {
+                if (protocols != SslProtocols.None &&
+                    CipherSuitesPolicyPal.WantsTls13(protocols))
                 {
-                    unsafe
-                    {
-                        Ssl.SslCtxSetVerify(innerContext, &VerifyClientCertificate);
-                    }
+                    protocols = protocols & (~SslProtocols.Tls13);
+                }
+            }
+            else if (CipherSuitesPolicyPal.WantsTls13(protocols) &&
+                        CipherSuitesPolicyPal.ShouldOptOutOfTls13(sslAuthenticationOptions.CipherSuitesPolicy, sslAuthenticationOptions.EncryptionPolicy))
+            {
+                if (protocols == SslProtocols.None)
+                {
+                    // we are using default settings but cipher suites policy says that TLS 1.3
+                    // is not compatible with our settings (i.e. we requested no encryption or disabled
+                    // all TLS 1.3 cipher suites)
+                    protocols = SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12;
```

Comment:
Is there no way to fetch what protocols None maps to?  We defer to the OS, right?  What happens if the OS decides that Tls is no longer on by default, for example?